DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,801 @@
+create copies). However, if no copy constructor in the overload set matches the exact
+qualified types of the source and the destination, the `alias this` is preferred (2).
+
+#### Generating copy constructors
+
+A copy constructor is generated for a `struct` S  if one of S's fields defines
+a copy constructor that S does not define.
+
+A field of a `struct` may define multiple copy constructors. In this situation,
+a copy constructor is generated for each overload.
+
+```d
+struct M
+{
+    this(ref immutable M) {}
+    this(ref immutable M) shared {}
+}
+
+struct S
+{
+    M m;
+    this(ref shared S) {}
+    /* The copy constructors with the following signatures are generated:
+        this(ref immutable S)
+        this(ref immutable S) shared
+     */
+}
+```
+
+The body of all the generated copy constructors does memberwise initialization for all
+the fields:
+
+```d
+{
+    this.field1 = s.field1;
+    this.field2 = s.field2;
+    ...;
+}
+```
+
+Inside the body of the generated copy constructors, for the fields that define a copy
+constructor, the assignment will be rewritten to a call to it; for those that do not,
+blitting is employed, if possible. If certain fields cannot perform copies
+(for example: the copy constructor is disabled), the generated copy constructor will be
+annotated with `@disable`.
+
+If copy constructors for fields intersect, a single copy constructor is generated:
+
+```d
+struct A
+{
+    int a;
+    this(ref A rhs)
+    {
+        this.a = rhs.a;
+    }
+}
+
+struct B
+{
+    int[] b;
+    this(ref B rhs)
+    {
+        this.b = rhs.b.dup;
+    }
+}
+
+struct C
+{
+    A a;
+    B b;
+    /* the following copy constructor is generated for C:
+
+    this(ref C rhs)
+    {
+        this.a = rhs.a;       // which, in turn, is lowered to a.__copyCtor(rhs.a);
+        this.b = rhs.b;       // which, in turn, is lowered to b.__copyCtor(rhs.b);
+    }
+    */
+}
+```
+
+#### Generating `opAssign` from a copy constructor
+
+The copy constructor is used to initialize an object from another object, whereas `opAssign` is
+used to copy an object to another object that has already been initialized:
+
+```d
+struct A
+{
+    int a;
+    immutable int id;
+    this(int a, int b)
+    {
+        this.a = a;
+        this.b = b;
+    }
+    this(ref A rhs)
+    {
+        this.a = rhs.a;
+        this.b = rhs.b;
+    }
+    void opAssign(A rhs)
+    {
+        this.a = rhs.a;
+    }
+}
+
+void main()
+{
+    A a = A(2);
+    A b = a;      // calls copy constructor;
+    b = a;        // calls opAssign;
+}
+```
+
+The reason why both the copy constructor and the `opAssign` method are needed is because
+the two are type checked differently: `opAssign` is type checked as a normal function whereas
+the copy constructor is type checked as a constructor (where the first assignment of non-mutable
+fields is allowed). However, in the majority of cases, the copy constructor body is identical
```

Comment:
However, in the majority of cases, the copy constructor body is identical to the `opAssign` one: -> In many cases, the body of the copy constructor will be identical to that of `opAssign`: